osf/utils/machines.py

```diff
@@ -143,3 +144,64 @@ def get_context(self):
             'provider_contact_email': self.machineable.provider.email_contact or 'contact@osf.io',
             'provider_support_email': self.machineable.provider.email_support or 'support@osf.io',
         }
+
+class RequestMachine(BaseMachine):
+    ActionClass = NodeRequestAction
+
+    def save_changes(self, ev):
+        """ Handles contributorship changes and state transitions
+        """
+        if ev.event.name == DefaultTriggers.ACCEPT.value:
+            self.machineable.target.add_contributor(self.machineable.creator, auth=Auth(ev.kwargs['user']), permissions=permissions.READ, send_email='{}_request'.format(self.machineable.request_type))
+        elif ev.event.name == DefaultTriggers.EDIT_COMMENT.value and self.action is not None:
+            self.machineable.comment = self.action.comment
+        self.machineable.save()
+
+    def resubmission_allowed(self, ev):
+        # TODO: [PRODUCT-395]
+        return False
+
+    def notify_submit(self, ev):
+        """ Notify admins that someone is requesting access
+        """
+        context = self.get_context()
+        context['contributors_url'] = '{}contributors/'.format(self.machineable.target.absolute_url)
+        context['project_settings_url'] = '{}settings/'.format(self.machineable.target.absolute_url)
+        for admin in self.machineable.target.admin_contributors:
+            mails.send_mail(
+                admin.username,
+                mails.ACCESS_REQUEST_SUBMITTED,
+                admin=admin,
+                **context
+            )
+
+    def notify_resubmit(self, ev):
+        """ Notify admins that someone is requesting access again
+        """
+        # TODO: [PRODUCT-395]
+        raise NotImplementedError()
+
+    def notify_accept_reject(self, ev):
+        """ Notify requester that admins have approved/denied
+        """
+        if ev.event.name == DefaultTriggers.REJECT.value:
+            context = self.get_context()
+            mails.send_mail(
+                self.machineable.creator.username,
+                mails.ACCESS_REQUEST_DENIED,
+                **context
+            )
+        else:
```

Comment:
The comment is nice, but if this isn't gonna handle other potential values of `DefaultTriggers` dunno if you need a whole empty `except` for it at the moment

Reply:
There are only two cases for possible event triggers here. One is covered by the above case, and the comment in this passed conditional branch is present to serve as an explanation for why the other case is not explicitly covered. The `if` is necessary, and `else` should add approximately 0.0ms to execution time. Also, it seemed to make sense to visually order it this way, as a comment at the top would need more words.

Reply:
all good!